framework/src/kernels/MatDiffusionBase.C

```diff
@@ -0,0 +1,163 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MatDiffusionBase.h"
+
+template <typename T, bool is_ad>
+InputParameters
+MatDiffusionBaseTempl<T, is_ad>::validParams()
+{
+  InputParameters params = GenericKernelGrad<is_ad>::validParams();
+
+  params.addParam<MaterialPropertyName>(
+      "diffusivity", "D", "The diffusivity value or material property");
+  params.addCoupledVar("args",
+                       "Optional vector of arguments for the diffusivity. If provided and "
+                       "diffusivity is a derivative parsed material, Jacobian contributions from "
+                       "the diffusivity will be automatically computed");
+  params.addCoupledVar("v",
+                       "Coupled concentration variable for kernel to operate on; if this "
+                       "is not specified, the kernel's nonlinear variable will be used as "
+                       "usual");
+  return params;
+}
+
+template <typename T, bool is_ad>
+MatDiffusionBaseTempl<T, is_ad>::MatDiffusionBaseTempl(const InputParameters & parameters)
+  : DerivativeMaterialInterface<JvarMapKernelInterface<GenericKernelGrad<is_ad>>>(parameters),
+    _diffusivity(this->template getGenericMaterialProperty<T, is_ad>("diffusivity")),
+    _ddiffusivity_dc(
+        is_ad ? nullptr
+              : &this->template getMaterialPropertyDerivative<T>("diffusivity", _var.name())),
+    _ddiffusivity_darg(!is_ad ? _coupled_moose_vars.size() : 0.0),
+    _is_coupled(isCoupled("v")),
+    _v_var(_is_coupled ? coupled("v") : _var.number()),
+    _grad_v(_is_coupled ? this->template coupledGenericGradient<is_ad>("v") : _grad_u)
+{
+  // fetch derivatives
+  for (unsigned int i = 0; i < _ddiffusivity_darg.size(); ++i)
+    _ddiffusivity_darg[i] = &this->template getMaterialPropertyDerivative<T>(
+        "diffusivity", _coupled_moose_vars[i]->name());
+}
+
+template <typename T, bool is_ad>
+void
+MatDiffusionBaseTempl<T, is_ad>::initialSetup()
+{
+  this->template validateNonlinearCoupling<T>("diffusivity");
+}
+
+template <>
+void
+MatDiffusionBaseTempl<Real, true>::initialSetup()
+{
+}
+
+template <>
+void
+MatDiffusionBaseTempl<RealTensorValue, true>::initialSetup()
+{
+}
+
+template <typename T, bool is_ad>
+GenericRealVectorValue<is_ad>
+MatDiffusionBaseTempl<T, is_ad>::precomputeQpResidual()
+{
+  return _diffusivity[_qp] * _grad_v[_qp];
+}
+
+template <>
+RealGradient
+MatDiffusionBaseTempl<Real, false>::computeQpCJacobian()
+{
+  return _diffusivity[_qp] * _grad_phi[_j][_qp];
+}
+
+template <>
+RealGradient
+MatDiffusionBaseTempl<RealTensorValue, false>::computeQpCJacobian()
+{
+  return _diffusivity[_qp] * _grad_phi[_j][_qp];
+}
+
+template <typename T, bool is_ad>
+RealGradient
+MatDiffusionBaseTempl<T, is_ad>::computeQpCJacobian()
+{
+  mooseError("Internal error, incorrect template specialization");
+}
```

Comment:
we have a better pattern for this! Have the non-AD version inherit form the generic template and add the methods.

Reply:
I'm trying, but a bit lost on this...do you have an example? Do you mean move `computeQpCJacobian` to `GenericKernelGrad`? If it's in the header, I have to have it for both ad and nonad right?

Reply:
or do you mean
```
MatDiffusionBaseTempl<T, is_ad>::computeQpCJacobian()
{
  if (is_ad)
    mooseError("Internal error, incorrect template specialization");
  do something...
}
```

Reply:
No. There should bo _no_ jacobian method in the AD class! You achieve that by having the non-AD class inherit from the AD template. Look at the Pressure BC for an example.

Reply:
ah I see!!

Reply:
That's a really slick way to do it! I can't get away from this issue though:
```
/Users/cmatthews/projects/moose/framework/src/kernels/MatDiffusionBase.C:67:10: error: use of undeclared identifier '_diffusivity'
  return _diffusivity[_qp] * _grad_phi[_j][_qp];
         ^
/Users/cmatthews/projects/moose/framework/src/kernels/MatDiffusionBase.C:67:23: error: use of undeclared identifier '_qp'
  return _diffusivity[_qp] * _grad_phi[_j][_qp];
  ```

Reply:
I thought this line should take care of it in the nonad: `using MatDiffusionBaseTempl<T, false>::MatDiffusionBaseTempl;`

Reply:
oh...i get it...lots more using required...

Reply:
You need a using for both `_diffusivity` and `_qp`

Reply:
I was just surprised it didn't pick up the `using` for `_qp` already.